### ISSUE TITLE
Ogre2.1 branch is long time update

### ISSUE THREAD

**Issue by 36yang** [MyGUI/mygui#219]
Can you update the ogre2.1 branch

**Comment by 36yang** [comment]
There are too many errors compile with Ogre2.2。 need support

**Comment by Altren** [comment]
I'm not familiar with Ogre2, so the only was I can help you is by telling what was the branch purpose. Someone implemented ogre2 render system and created a branch for it, but since it never was finished it haven't got into master, thus it is outdated now. The plan was to add Ogre2 render system in master, and keep it as well as Ogre1.
May be @al2950 , @PieroV or scrawl (hm, scrawl user name is gone) can help you with that.

**Comment by al2950** [comment]
Hi

The issue is Ogre 2+ has changed significantly between minor releases making it very difficult to create a single plugin that supports 2.0, 2.1, 2.2 & 2.3, and so this branch has gone a bit stale. Ogre 2+ has become more stable in the last year or so, so it might be worth revisiting this and creating a new branch and starting again making it support Ogre 2.2++. I have a my own version of mygui that compiles with Ogre 2.2 but I would not say its fully functional.

**Comment by Altren** [comment]
Would it be OK to support only latest Ogre2 version? Having more that one Ogre2 render system sounds crazy and adds too much hussle to maintain that.
Or if API is not changing too rapidly may be we can have one version with a few `#if OGRE_VERSION` checks, that makes it work with all 2.2+ versions.

**Comment by sercero** [comment]
Also, it might make sense to have Ogre2 as an additional render system instead of a separate branch.

**Comment by al2950** [comment]
> Would it be OK to support only latest Ogre2 version? Having more that one Ogre2 render system sounds crazy and adds too much hussle to maintain that.
> Or if API is not changing too rapidly may be we can have one version with a few `#if OGRE_VERSION` checks, that makes it work with all 2.2+ versions.

Yes... But please note I started to write code to support all Ogre 2.x versions with #if OGRE_VERSION but it got too messy, so it would have to just be 2.2+

> Also, it might make sense to have Ogre2 as an additional render system instead of a separate branch.
This branch does actually create an additional render system

**Comment by 36yang** [comment]
Yes, Only Support the lasted 2.2+ is OK, I can  wait for it.

**Comment by cryham** [comment]
It would be great to have official working MyGUI with latest Ogre 2.2+.

**Comment by cryham** [comment]
Just for info, I have a working fork of MyGUI for latest Ogre-Next 3.0
(Windows CMake not fixed yet)
https://github.com/cryham/mygui-next/commits/ogre3
Ogre topic with details here: https://forums.ogre3d.org/viewtopic.php?t=96874

**Comment by Altren** [comment]
I'm willing to accept MR, that would be compatible with master. I recommend rebasing to it, because there is at least one major feature, related to render systems (custom shaders support), that also require some extra code in your branch. Ogre 2 branch was very old.

**Comment by cryham** [comment]
Yeah, after this commit https://github.com/MyGUI/mygui/commit/e19a011076024ad0c05e8413e6129879099216e8 it will be rather impossible to upmerge changes (from any forks).
I think the only way would be to apply fork changes (like our for Ogre-Next) again to the new master of MyGui. Probably means more work. (I still am very far from considering doing it, busy with other stuff).

**Comment by Apteryks** [comment]
@cryham hello! I'm currently packaging Stunt Rally 3 for GNU Guix, and having a single MyGUI package would be nice. If it's fallen into cracks, I'd encourage you to consider upstreaming your changes again... it's less work down the road.

**Comment by Altren** [comment]
[PR with OgreNext platform](https://github.com/MyGUI/mygui/pull/313). It would be great if someone can review it

**Comment by Altren** [comment]
Ok, all tests passed, Demos, Tools, custom shaders and everything else works.
I'm going to merge this, since it might take a while for someone to try this. And it would not hurt having this in master.
Any reviews (even only for architecture decisions) are highly appreciated.
